### PR TITLE
EffectiveTldFinder to parse Internationalized Domain Names (IDN)

### LINKED ISSUE #179: EffectiveTldFinder to parse IDNs
Unicode [Internationalized Domain Names](https://en.wikipedia.org/wiki/Internationalized_domain_name) in the are not processed by EffectiveTldFinder, domain name and TLD are only extracted for the Punycode representation. Also hostnames with mixed usage of Unicode and Punycode, e.g. [xn--90a1af.бесплатныеобъявления.рф](http://xn--90a1af.бесплатныеобъявления.рф/). See also #172 which added already unit tests (commented out) for this issue.

### PR DESCRIPTION
**Pull request by sebastian-nagel**
fixes #179
- allow eTLDs to appear as IDN variants
  * one or more dot-separated parts as IDN,
  * optionally others as punycoded IDNA
- speed-up lookup of public suffixes / eTLDs using a suffix trie
- update public suffix list to recent version
- add unit tests for wildcard eTLDs

### CODE REVIEW

File: src/main/java/crawlercommons/domains/EffectiveTldFinder.java [hunk 1]
```diff
@@ -262,43 +286,38 @@ public static String getAssignedDomain(String hostname, boolean strict) {
...
         }
         // clip hostname one dot-separated element before eTLD
-        if (hostname.endsWith(etld.getDomain())) {
-            int etldStartPos = hostname.length() - etld.getDomain().length() - 1;
-            if (hostname.charAt(etldStartPos) != DOT) {
-                // should not happen: no dot before TLD
+        int etldStartPos = res.offset - 1;
+        if (hostname.charAt(etldStartPos) != DOT) {
+            // should not happen: no dot before TLD
+            return (strict ? null : hostname);
```
Comment (kkrugler) [inline src/main/java/crawlercommons/domains/EffectiveTldFinder.java]: debug logging might be helpful here, especially for the non-strict case.

File: src/main/java/crawlercommons/domains/SuffixTrie.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
+package crawlercommons.domains;
```
Comment (kkrugler) [inline src/main/java/crawlercommons/domains/SuffixTrie.java]: Copyright? Where did this code come from?

File: src/test/java/crawlercommons/domains/EffectiveTldFinderTest.java [hunk 3]
```diff
@@ -104,6 +110,14 @@ public final void testExceptionEtld() throws Exception {
         assertEquals("city.kawasaki.jp", etld.getDomain());
         assertEquals("city.kawasaki.jp", EffectiveTldFinder.getAssignedDomain("www.city.kawasaki.jp"));
         assertEquals(".kawasaki.jp", EffectiveTldFinder.getAssignedDomain(".kawasaki.jp"));
+        assertNull(EffectiveTldFinder.getAssignedDomain(".kawasaki.jp", true));
+        assertEquals("city.kawasaki.jp", EffectiveTldFinder.getAssignedDomain("city.kawasaki.jp", true));
+        // an wildcard eTLD itself is not a valid domain
```
Comment (kkrugler) [inline src/test/java/crawlercommons/domains/EffectiveTldFinderTest.java]: "a wildcard"

**Comment by sebastian-nagel** [comment]
Hi @kkrugler, thanks for the review. I've made the suggested changes.

**Comment by kkrugler** [comment]
Oh, is there an update to CHANGES.txt?

**Comment by sebastian-nagel** [comment]
Also updated CHANGES.txt. Thanks!

**Comment by kkrugler** [comment]
Thanks @sebastian-nagel!

**Pull request merged**